tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/graal/phases/TornadoHalfFloatVectorOffset.java

```diff
@@ -0,0 +1,146 @@
+import uk.ac.manchester.tornado.api.types.arrays.TornadoNativeArray;
+import uk.ac.manchester.tornado.drivers.spirv.graal.nodes.vector.SPIRVVectorValueNode;
+import uk.ac.manchester.tornado.runtime.graal.nodes.VectorHalfRead;
+
+import java.util.Optional;
+
+/**
+ * This compiler phase ensures that the elements of half-float vectors are being
+ * accessed using the correct offsets. This is essential because the JavaKind of
+ * the half type is declared as Object in order to avoid issues with the Stamp during
+ * the sketching.
+ */
+public class TornadoHalfFloatVectorOffset extends Phase {
+
+    private static int HALF_SIZE = 2;
+    private static long HEADER_SIZE = TornadoNativeArray.ARRAY_HEADER;
+
+    @Override
+    public Optional<NotApplicable> notApplicableTo(GraphState graphState) {
+        return ALWAYS_APPLICABLE;
+    }
+
+    protected void run(StructuredGraph graph) {
+        for (ReadNode readNode : graph.getNodes().filter(ReadNode.class)) {
+            if (readNode.successors().filter(VectorHalfRead.class).isNotEmpty()) {
+                VectorHalfRead vectorHalfRead = readNode.successors().filter(VectorHalfRead.class).first();
+                replaceReadOffset(readNode, vectorHalfRead, graph);
+                deleteFixed(vectorHalfRead);
+            }
+        }
+
+        for (WriteNode writeNode : graph.getNodes().filter(WriteNode.class)) {
+            if (writeNode.value() instanceof SPIRVVectorValueNode && ((SPIRVVectorValueNode) writeNode.value()).getSPIRVKind().isHalf()) {
+                replaceWriteOffset(writeNode, graph);
+            }
+        }
+
+    }
+
+    private static void replaceReadOffset(ReadNode readNode, VectorHalfRead vectorHalfRead, StructuredGraph graph) {
+        OffsetAddressNode ptxAddressNode = (OffsetAddressNode) readNode.getAddress();
+        ValueNode index = ptxAddressNode.getOffset();
+        if (index instanceof ConstantNode) {
+            int vectorReadIndex = vectorHalfRead.getIndex();
+            // if the index value has been initialized
+            if (vectorReadIndex != -1) {
+                Constant shortOffset = new RawConstant(vectorReadIndex * HALF_SIZE + HEADER_SIZE);
+                ConstantNode shortOffsetNode = new ConstantNode(shortOffset, StampFactory.forKind(JavaKind.Int));
+                graph.addWithoutUnique(shortOffsetNode);
+                ptxAddressNode.replaceFirstInput(index, shortOffsetNode);
+                if (index.usages().isEmpty()) {
+                    index.safeDelete();
+                }
+            }
+        } else if (index.inputs().filter(LeftShiftNode.class).isNotEmpty()) {
+            LeftShiftNode leftShiftNode = index.inputs().filter(LeftShiftNode.class).first();
+            ConstantNode currentOffset = leftShiftNode.inputs().filter(ConstantNode.class).first();
+            // if the shifting is by 3 (for float values)
+            if (currentOffset.getValue().toValueString().equals("3")) {
```

Comment:
Why the shift is by 3 for a float value? Can we generalize this?

Reply:
The comment above is wrong, it's not because of float types, it's because the JavaKind for half is Object (8 bytes). This was done because otherwise we were having issues with the stamp. I will update the comment to reflect that.